File: src/server/src/main/java/com/project/domain/workout/WorkoutExercise.java
package com.project.domain.workout;

import com.project.domain.BaseEntity;
import com.project.utilities.FindUtil;
import com.project.utilities.Time;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.function.Predicate;

@Entity(name = "workout_exercises")
public class WorkoutExercise extends BaseEntity {

    @Id
    private final UUID id;

    private UUID exerciseId;

    private String name;

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true, mappedBy = "exercise")
    private final List<WorkoutSet> sets;

    @ManyToOne(fetch = FetchType.LAZY)
    private WorkoutTemplate workoutTemplate;

    @ManyToOne(fetch = FetchType.LAZY)
    private OriginalWorkoutTemplate originalWorkoutTemplate;

    protected WorkoutExercise() {
        super();
        this.id = UUID.randomUUID();
        this.sets = new ArrayList<>();
    }

    protected WorkoutExercise(Exercise exercise) {
        this();
        this.fromExercise(exercise);
        this.sets.add(WorkoutSet.from(0, 10));
        this.sets.get(0).setExercise(this);
    }

    protected WorkoutExercise(Exercise exercise, List<WorkoutSet> sets) {
        this();
        this.fromExercise(exercise);
        this.setSets(sets);
    }

    protected WorkoutExercise(Exercise exercise, OriginalWorkoutTemplate workoutTemplate, List<WorkoutSet> sets) {
        this();
        this.fromExercise(exercise);
        this.originalWorkoutTemplate = workoutTemplate;
        this.setSets(sets);
    }

    protected WorkoutExercise(UUID exerciseId, WorkoutTemplate workoutTemplate, List<WorkoutSet> sets) {
        this();
        this.exerciseId = exerciseId;
        this.workoutTemplate = workoutTemplate;
        this.setSets(sets);
    }

    public UUID getId() {
        return this.id;
    }

    public UUID getExerciseId() {
        return this.exerciseId;
    }

    public String getName() {
        return this.name;
    }

    public List<WorkoutSet> getSets() {
        return this.sets;
    }

    public WorkoutTemplate getWorkoutTemplate() {
        return this.workoutTemplate;
    }

    public OriginalWorkoutTemplate getOriginalWorkoutTemplate() {
        return this.originalWorkoutTemplate;
    }

    public void setWorkoutTemplate(WorkoutTemplate workoutTemplate) {
        this.workoutTemplate = workoutTemplate;
        this.setModifiedAt(Time.utcNow());
    }

    public static WorkoutExercise from(Exercise exercise) {
        return new WorkoutExercise(exercise);
    }

    public static WorkoutExercise from(Exercise exercise, List<WorkoutSet> sets) {
        return new WorkoutExercise(exercise, sets);
    }

    public static WorkoutExercise from(Exercise exercise, WorkoutTemplate workoutTemplate, List<WorkoutSet> sets) {
        return new WorkoutExercise(exercise.getId(), workoutTemplate, sets);
    }

    public static WorkoutExercise from(Exercise exercise, OriginalWorkoutTemplate workoutTemplate, List<WorkoutSet> sets) {
        return new WorkoutExercise(exercise, workoutTemplate, sets);
    }

    public void updateSet(UUID setId, double weight, int repetitions) {
        Predicate<WorkoutSet> predicate = ws -> ws.getId().equals(setId);
        var set = FindUtil.findByProperty(this.sets, predicate);
        set.ifPresent(s -> {
            s.setWeight(weight);
            s.setRepetitions(repetitions);
        });
    }

    public void addSet(int repetitions, double weight) {
        var set = WorkoutSet.from(weight, repetitions);
        set.setExercise(this);
        this.sets.add(set);
        this.setModifiedAt(Time.utcNow());
    }

    private void fromExercise(Exercise exercise) {
        this.exerciseId = exercise.getId();
        this.name = exercise.getName();
    }

    private void setSets(List<WorkoutSet> sets) {
        sets.forEach(s -> s.setExercise(this));
        this.sets.addAll(sets);
    }
}

File: src/server/src/main/java/com/project/domain/workout/WorkoutExercise.kt
package com.project.domain.workout

import com.project.domain.BaseEntity
import com.project.utilities.FindUtil.findByProperty
import com.project.utilities.Time.utcNow
import jakarta.persistence.CascadeType
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.Id
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import java.util.UUID

@Entity(name = "workout_exercises")
class WorkoutExercise protected constructor() : BaseEntity() {
    @Id
    val id: UUID = UUID.randomUUID()

    var exerciseId: UUID? = null
        private set

    var name: String? = null
        private set

    @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true, mappedBy = "exercise")
    val sets: MutableList<WorkoutSet> = ArrayList()

    @ManyToOne(fetch = FetchType.LAZY)
    var workoutTemplate: WorkoutTemplate? = null

    @ManyToOne(fetch = FetchType.LAZY)
    var originalWorkoutTemplate: OriginalWorkoutTemplate? = null
        private set

    protected constructor(exercise: Exercise) : this() {
        this.fromExercise(exercise)
        sets.add(WorkoutSet.from(0.0, 10))
        sets[0].exercise = this
    }

    protected constructor(exercise: Exercise, sets: List<WorkoutSet>) : this() {
        this.fromExercise(exercise)
        this.addSets(sets)
    }

    protected constructor(exercise: Exercise, workoutTemplate: OriginalWorkoutTemplate?, sets: List<WorkoutSet>) : this() {
        this.fromExercise(exercise)
        this.originalWorkoutTemplate = workoutTemplate
        this.addSets(sets)
    }

    protected constructor(exerciseId: UUID?, workoutTemplate: WorkoutTemplate?, sets: List<WorkoutSet>) : this() {
        this.exerciseId = exerciseId
        this.workoutTemplate = workoutTemplate
        this.addSets(sets)
    }

    fun updateSet(setId: UUID, weight: Double, repetitions: Int) {
        val predicate = { ws: WorkoutSet? -> ws?.id == setId }
        val set = findByProperty(this.sets, predicate)
        set?.repetitions = repetitions
        set?.setWeight(weight)
    }

    fun addSet(repetitions: Int, weight: Double) {
        val set: WorkoutSet = WorkoutSet.Companion.from(weight, repetitions)
        set.exercise = this
        sets.add(set)
        this.modifiedAt = utcNow()
    }

    private fun fromExercise(exercise: Exercise) {
        this.exerciseId = exercise.id
        this.name = exercise.name
    }

    private fun addSets(sets: List<WorkoutSet>) {
        sets.forEach { s: WorkoutSet -> s.exercise = this }
        this.sets.addAll(sets)
    }

    companion object {
        @JvmStatic
        fun from(exercise: Exercise): WorkoutExercise {
            return WorkoutExercise(exercise)
        }

        fun from(exercise: Exercise, sets: List<WorkoutSet>): WorkoutExercise {
            return WorkoutExercise(exercise, sets)
        }

        fun from(exercise: Exercise, workoutTemplate: WorkoutTemplate?, sets: List<WorkoutSet>): WorkoutExercise {
            return WorkoutExercise(exercise.id, workoutTemplate, sets)
        }

        fun from(exercise: Exercise, workoutTemplate: OriginalWorkoutTemplate?, sets: List<WorkoutSet>): WorkoutExercise {
            return WorkoutExercise(exercise, workoutTemplate, sets)
        }
    }
}
